typed/merge.go

```diff
@@ -280,72 +293,72 @@ func (w *mergingWalker) doList(t *schema.List) (errs ValidationErrors) {
 	return errs
 }
 
-func (w *mergingWalker) visitMapItems(t *schema.Map, lhs, rhs *value.Map) (errs ValidationErrors) {
-	out := &value.Map{}
+func (w *mergingWalker) visitMapItem(t *schema.Map, out map[string]interface{}, key string, lhs, rhs value.Value) (errs ValidationErrors) {
+	fieldType := t.ElementType
+	if sf, ok := t.FindField(key); ok {
+		fieldType = sf.Type
+	}
+	pe := fieldpath.PathElement{FieldName: &key}
+	w2 := w.prepareDescent(pe, fieldType)
+	w2.lhs = lhs
+	w2.rhs = rhs
+	if newErrs := w2.merge(); len(newErrs) > 0 {
+		errs = append(errs, newErrs.WithPrefix(pe.String())...)
+	} else if w2.out != nil {
+		out[key] = *w2.out
+	}
+	w.finishDescent(w2)
+	return nil
+}
+
+func (w *mergingWalker) visitMapItems(t *schema.Map, lhs, rhs value.Map) (errs ValidationErrors) {
+	out := map[string]interface{}{}
```

Comment:
Hm, I'm really sad about the reduction in type safety.

I think for the reflect based one to work this is going to have to switch to a factory and/or we're going to have to add getters/setters to the value interface :/

Reply:
Not disagreeing. We do plan on improving that.

Reply:
I suspect this change is big and old enough already though ... and blocking a couple of people :-/

Reply:
not asking for a change here :'(